Support drag & drop of a profile into already-loaded profiler (Fixes #2001)

brisad opened the pull request:
Based on the work in https://github.com/firefox-devtools/profiler/pull/2072 Fixes #2001 

I didn't do anything on the back-button problem, since the same problem is already there even without this change. I thought I could take a look at it in a separate issue if that is OK?

I would have liked to have a better component test, but didn't manage to get a test for the actual dropping of a file. See the comment in the new test.

gregtatum commented on src/test/components/DragAndDrop.test.js:
```diff
@@ -0,0 +1,46 @@
+    // `retrieveProfileFromFile`, but jsdom is not supporting
+    // `dataTransfer`. See
+    // https://github.com/testing-library/react-testing-library/issues/339
+    // for more info.
```
Thanks, I filed #2366. It might be good to reference that here as well.

gregtatum commented on src/test/components/DragAndDrop.test.js:
```diff
@@ -0,0 +1,46 @@
+import { render, fireEvent } from 'react-testing-library';
+import DragAndDrop from '../../components/app/DragAndDrop';
+
+describe('app/DragAndDrop', () => {
```
Thanks for the tests!

gregtatum commented on src/reducers/app.js:
```diff
@@ -33,6 +33,7 @@ const view: Reducer<AppViewState> = (
       return { phase: 'FATAL_ERROR', error: action.error };
     case 'WAITING_FOR_PROFILE_FROM_ADDON':
     case 'WAITING_FOR_PROFILE_FROM_URL':
+    case 'WAITING_FOR_PROFILE_FROM_FILE':
```
Is this view introduced from this PR, or did we just miss it?

brisad replied:
No, this was already existing.

gregtatum commented on src/components/app/DragAndDrop.css:
```diff
@@ -0,0 +1,36 @@
+  isolation: isolate;
+}
+
+.dropMessageWrapper {
```
It'd be good to make these classes consistent in naming to avoid future naming conflicts. We usually use the component name to make up the names, so:

```
.dragAndDropMessageWrapper {}
.dragAndDropMessageWrapper.dragging {}
.dragAndDropMessage {}
```

brisad commented:
Thanks for the review! I have addressed your comments and I've got something reviewable again.

I have connected the component to the Redux store, and it is mounted at the highest point in `Root.js`. This gets us drag-and-drop functionality at more or less all pages in the profiler. I have blacklisted a few states where the profiler is loading so that we aren't able to drag and drop during that phase.

I've restored the visual appearance on the home-page, so that the overlay only covers the smaller rectangle, like it did before. The zip-file viewer similarly has a smaller overlay. On other pages, like the error page, the overlay spans the whole screen. I thought that looked the nicest, but it can be customized by using the new `<DragAndDropOverlay>` component.

The bug of not resetting the reducer state has also been solved now. Good catch!

There are a few more things that I'd like to address, listed below, but I thought I'd ask for a review first to see whether my approach of registering a custom overlay is an okay one.

Oh and please ignore the merge conflict. I'll address that later.

Thanks!

---

- [ ] Drag and drop spamming
Dragging a file dispatches a bunch of `START_DRAGGING`/`STOP_DRAGGING` actions. I don't know if this is a problem, but it doesn't look too nice in the console.
- [ ] Block scrolling behind overlay
As I drag a file over the overlay in the profile view, I can move it over the thread list and make it scroll, behind the overlay. I've tried to fix this, but I've been unsuccessful so far.
- [ ] Throw error if there's more than 1 overlay registered
If there are multiple overlays registered, we should throw an error instead of silently allowing them.

gregtatum commented on src/actions/app.js:
```diff
@@ -151,3 +151,19 @@ export function reportTrackThreadHeight(
   return { type: 'DISMISS_NEWLY_PUBLISHED' };
 }
+
+export function startDragging(): Action {
```
I generally recommend adding code comments to the start of all the action creators, as these are the public API that are consumed in the app.

gregtatum commented on src/components/app/DragAndDrop.css:
```diff
@@ -0,0 +1,38 @@
+.dragAndDropArea {
```
I think some of the CSS files are lacking the license preamble, but we should add that to this file.:

```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

gregtatum commented on src/components/app/DragAndDrop.js:
```diff
@@ -0,0 +1,222 @@
+          dragAndDropArea div creates its own stacking context, so
+          even if it contains children with high z-indexes, the
+          default overlay will still appear on top when shown.*/
+        this.props.useDefaultOverlay && <DragAndDropOverlay isDefault={true} />}
```
Please use a ternary here instead of the `&&` switching. I find the `&&` and `||` behaviors take a lot more time to reason with what actually gets passed in, rather than the arm branching of a ternary.

`useDefaultOverlay ? <DragAndDropOverlay isDefault={true} />} : null`

gregtatum commented on src/selectors/app.js:
```diff
@@ -164,3 +171,17 @@ export const getTimelineHeight: Selector<null | CssPixels> = createSelector(
   }
 );
+
+export const getIsNewProfileLoadAllowed: Selector<boolean> = createSelector(
```
Selectors are also really good places for code documentation, as they are the public interface into the data. It would probably be good to explain this mechanism a little bit.

gregtatum commented on src/reducers/app.js:
```diff
@@ -183,6 +184,31 @@ const isNewlyPublished: Reducer<boolean> = (state = false, action) => {
   }
 };
 
+const isDragAndDropDragging: Reducer<boolean> = (state = false, action) => {
```
Reducers are also a good place for some light code documentation. See the reducer above as an example.

brisad commented:
Thanks for the review! I've addressed your comments and written some documentation. :-)

brisad commented:
I will follow up the remaining issues in separate PRs/issues.
